File: yami4/src/python3/examples/subscription/publisher.py
import random
import sys
import time
import yami

if len(sys.argv) != 2:
    print("expecting one parameter: publisher destination")
    exit()

publisher_address = sys.argv[1]

try:
    random_value = yami.ValuePublisher()

    with yami.Agent() as publisher_agent:
        resolved_address = publisher_agent.add_listener(
            publisher_address)

        print("The publisher is listening on", resolved_address)

        publisher_agent.register_value_publisher(
            "random_number", random_value)

        # publish random numbers forever
        while True:
            rnd = random.randint(0, 99)
            content = {"value":rnd}

            print("publishing value", rnd)

            random_value.publish(content)

            time.sleep(1)

except Exception as e:
    print("error:", e)

File: yami4/src/objc/examples/subscription/publisher.m
#import <yami4-objc/YAMI4.h>

#import <Foundation/NSAutoreleasePool.h>
#import <Foundation/NSString.h>
#import <Foundation/NSThread.h>

#include <stdio.h>
#include <stdlib.h>


int main(int argc, char * argv[])
{
    if (argc != 2)
    {
        puts("expecting one parameter: publisher destination");
        return -1;
    }

    NSAutoreleasePool * pool =
        [[NSAutoreleasePool alloc] init];
    
    NSString * publisherAddress =
        [[NSString alloc] initWithUTF8String:argv[1]];

    @try
    {
        YAMI4ValuePublisher * randomValue =
            [YAMI4ValuePublisher new];
        
        YAMI4Agent * publisherAgent = [YAMI4Agent new];

        NSString * resolvedAddress =
            [publisherAgent addListener:publisherAddress];

        printf("The publisher is listening on %s\n",
            [resolvedAddress UTF8String]);
        
        [publisherAgent registerValuePublisher:@"random_number"
            publisher:randomValue];

        // publish random numbers forever
        YAMI4Parameters * content = [YAMI4Parameters new];
        while (YES)
        {
            int random = rand() % 100;
            [content setInteger:@"value" value:random];
            
            printf("publishing value %d\n", random);
            
            [randomValue publish:content];
            
            // pause for 1s
            [NSThread sleepForTimeInterval:1.0];
        }
    }
    @catch (NSException * e)
    {
        printf("error: %s\n", [[e reason] UTF8String]);
    }
    
    [pool drain];
    
    return 0;
}

